web/src/message_notifications.js

```diff
@@ -0,0 +1,447 @@
+        key = message.display_reply_to;
+    } else {
+        const stream_name = stream_data.get_stream_name_from_id(message.stream_id);
+        key = message.sender_full_name + " to " + stream_name + " > " + message.topic;
+    }
+
+    return key;
+}
+
+function remove_sender_from_list_of_recipients(message) {
+    return `, ${message.display_reply_to}, `
+        .replace(`, ${message.sender_full_name}, `, ", ")
+        .slice(", ".length, -", ".length);
+}
+
+function get_notification_title(message, content, msg_count) {
+    let title = message.sender_full_name;
+    let other_recipients;
+
+    if (msg_count > 1) {
+        title = msg_count + " messages from " + title;
+    }
+
+    switch (message.type) {
+        case "test-notification":
+            other_recipients = remove_sender_from_list_of_recipients(message);
+            break;
+        case "private":
+            other_recipients = remove_sender_from_list_of_recipients(message);
+            if (message.display_recipient.length > 2) {
+                // If the message has too many recipients to list them all...
+                if (content.length + title.length + other_recipients.length > 230) {
+                    // Then count how many people are in the conversation and summarize
+                    // by saying the conversation is with "you and [number] other people"
+                    other_recipients =
+                        other_recipients.replaceAll(/[^,]/g, "").length + " other people";
+                }
+
+                title += " (to you and " + other_recipients + ")";
+            } else {
+                title += " (to you)";
+            }
+            break;
+        case "stream": {
+            const stream_name = stream_data.get_stream_name_from_id(message.stream_id);
+            title += " (to " + stream_name + " > " + message.topic + ")";
+            break;
+        }
+    }
+
+    return title;
+}
+
+export function process_notification(notification) {
+    const message = notification.message;
+    const content = get_notification_content(message);
+    const key = get_notification_key(message);
+    let notification_object;
+    let msg_count = 1;
+
+    debug_notification_source_value(message);
+
+    if (notifications.notice_memory.has(key)) {
+        msg_count = notifications.notice_memory.get(key).msg_count + 1;
+        notification_object = notifications.notice_memory.get(key).obj;
+        notification_object.close();
+    }
+
+    const title = get_notification_title(message, content, msg_count);
+
+    if (notification.desktop_notify) {
+        const icon_url = people.small_avatar_url(message);
+        notification_object = new notifications.NotificationAPI(title, {
+            icon: icon_url,
+            body: content,
+            tag: message.id,
+        });
+        notifications.notice_memory.set(key, {
+            obj: notification_object,
+            msg_count,
+            message_id: message.id,
+        });
```

Comment:
This is fine for now but I have to assume we'd want to have `notice_memory` stop being exported to have a clear separation of purposes for the modules -- the `notifications.js` module (which maybe we'd rename to `desktop_notifications.js` once we're done with this) would have a bit of an API for the rest of the app, mostly this module, to talk to.

But I don't think it's worth blocking on such a refactor.